File: csharp/perfect-numbers/PerfectNumbers.cs
using System;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0)
        {
            throw new ArgumentOutOfRangeException("Not natural number");
        }
        int dividerSum = DividerSum(number);
        if (dividerSum < number)
        {
            return Classification.Deficient;
        }
        if (dividerSum == number)
        {
            return Classification.Perfect;
        }
        return Classification.Abundant;

    }

    public static int DividerSum(int number)
    {
        int sum = 0;
        for (int i = 1; i <= number / 2; i++)
        {
            if (number % i == 0)
            {
                sum += i;
            }
        }
        return sum;
    }
}

File: cpp/perfect-numbers/perfect_numbers.cpp
#include "perfect_numbers.h"
#include <stdexcept>

namespace perfect_numbers
{

    classification classify(int number)
    {
        if (number <= 0)
        {
            throw std::domain_error("Not natural number");
        }

        int divisor_sum = 1;

        for (int i = 2; i * i <= number; ++i)
        {
            if (number % i == 0)
            {
                divisor_sum += i;

                if (i != number / i)
                {
                    divisor_sum += number / i;
                }
            }
        }

        if (number == 1)
        {
            divisor_sum = 0;
        }

        if (divisor_sum < number)
        {
            return classification::deficient;
        }
        else if (divisor_sum > number)
        {
            return classification::abundant;
        }

        return classification::perfect;
    }

} // namespace perfect_numbers
